solver/src/naive_solver/multi_order_solver.rs

```diff
@@ -0,0 +1,437 @@
+#![allow(dead_code)]
+
+use super::single_pair_settlement::{AmmSwapExactTokensForTokens, SinglePairSettlement};
+use crate::settlement::Trade;
+use anyhow::{anyhow, Result};
+use model::order::{OrderCreation, OrderKind};
+use num::{bigint::Sign, BigInt};
+use std::collections::HashMap;
+use web3::types::{Address, U256};
+
+#[derive(Default, Debug)]
+struct TokenContext {
+    address: Address,
+    reserve: U256,
+    buy_volume: U256,
+    sell_volume: U256,
+}
+
+impl TokenContext {
+    pub fn is_excess_after_fees(&self, deficit: &TokenContext) -> bool {
+        1000 * u256_to_bigint(&self.reserve)
+            * (u256_to_bigint(&deficit.sell_volume) - u256_to_bigint(&deficit.buy_volume))
+            < 997
+                * u256_to_bigint(&deficit.reserve)
+                * (u256_to_bigint(&self.sell_volume) - u256_to_bigint(&self.buy_volume))
+    }
+}
+
+///
+/// Computes a settlement using orders of a single pair and the direct AMM between those tokens.
+/// Panics if orders are not already filtered for a specific token pair, or the reserve information
+/// for that pair is not available.
+///
+pub fn solve(
+    orders: impl Iterator<Item = OrderCreation> + Clone,
+    reserves: &HashMap<Address, U256>,
+) -> SinglePairSettlement {
+    let (context_a, context_b) = split_into_contexts(orders.clone(), reserves);
+    if context_a.is_excess_after_fees(&context_b) {
+        solve_with_uniswap(orders, &context_b, &context_a)
+    } else if context_b.is_excess_after_fees(&context_a) {
+        solve_with_uniswap(orders, &context_a, &context_b)
+    } else {
+        solve_without_uniswap(orders, &context_a, &context_b)
+    }
+}
+
+///
+/// Creates a solution using the current AMM spot price, without using any of its liquidity
+///
+fn solve_without_uniswap(
+    orders: impl Iterator<Item = OrderCreation> + Clone,
+    context_a: &TokenContext,
+    context_b: &TokenContext,
+) -> SinglePairSettlement {
+    SinglePairSettlement {
+        clearing_prices: maplit::hashmap! {
+            context_a.address => context_b.reserve,
+            context_b.address => context_a.reserve,
+        },
```

Comment:
In the future, this should probably not be so naive as a single +/- 1 atom can cause a 0.3% jump in price. Absolutely not important for this PR though.